### PR TITLE
Fix desugaring bug

### LINKED ISSUE #275: Module grouping messes up order of imports
Imports (for example function imports) can be defined either as `(func ... (import ...))` or `(import ... (func ...))`.
The former should just desugar to the latter, without changing the order of imports.
However, the way KWasm first groups the different definitions in a module, and then desugars them, the above imports would be reordered, changing their indices.

The following file works in the reference interpreter, but not in KWasm, where it results in `0`.

```
(module
  (func (export "foo") (result i32) (i32.const 0))
  (func (export "bar") (result i32) (i32.const 1))
)
(register "a")

(module
(func (import "a" "bar") (result i32))
(import "a" "foo" (func (result i32)))
(export "bar" (func 0) )
)

(assert_return (invoke "bar") (i32.const 1))
```

See branch
<https://github.com/kframework/wasm-semantics/tree/desugaring>

### LINKED ISSUE #278: Change as many desugarings as possible in WASM-TEXT to be macros
This brings us closer to the official spec, and also avoids coordination mistakes such as #275.

### PR DESCRIPTION
**Pull request by hjorthjort**
Fixes #275, closes #278.

The specification (and reference interpreter supports this) says that all the abbreviations (syntactic sugars) in the text format are desugared immediately, and only after are the different definitions grouped. By not desugaring before grouping, we caused the order of imports, data declarations and elem declarations to get changed. This turns those desugarings into macros, so that they get applied before grouping.

### CODE REVIEW

**Comment by hjorthjort** [comment]
This PR is stuck on this bug:

https://github.com/kframework/k/issues/1000

The issue is when there is a hex const after a `table` or `memory` inlined with `elem` or `data` such as the `"break-inner"` function in `block.wast`.

File: tests/simple/modules.wast [hunk 1]
```diff
@@ -55,4 +55,24 @@
...
+    (local.set 0
+      (i32.add (local.get 0) (block (result i32) (i32.ctz (br 0 (i32.const 0x4)))))
+    )
+    (local.set 0
+      (i32.add (local.get 0) (block (result i32) (i32.ctz (block (result i32) (br 1 (i32.const 0x8))))))
+    )
+    (local.get 0)
+  )
+)
+
```
Comment (hjorthjort) [inline tests/simple/modules.wast]: Move this to the `desugaring.wast` test

File: tests/simple/desugaring.wast [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  (table funcref (elem 1))
+  (func (result i32) (i32.const 0))
+  (func (result i32) (i32.const 1))
+  (func (export "biz") (result i32)
+    (call_indirect (result i32) (i32.const 0))
+  )
+)
+
+(assert_return (invoke "biz") (i32.const 1))
+
```
Comment (hjorthjort) [inline tests/simple/desugaring.wast]:
```suggestion

;; Regression test: A module with hex integers inside a function after a `table` with inline `elem`.
(module
  (table funcref (elem))
  (func (export "break-inner") (result i32)
    (local i32)
    (local.set 0 (i32.const 0))
    (local.set 0 (i32.add (local.get 0) (block (result i32) (block (result i32) (br 1 (i32.const 0x1))))))
    (local.set 0 (i32.add (local.get 0) (block (result i32) (block (br 0)) (i32.const 0x2))))
    (local.set 0
      (i32.add (local.get 0) (block (result i32) (i32.ctz (br 0 (i32.const 0x4)))))
    )
    (local.set 0
      (i32.add (local.get 0) (block (result i32) (i32.ctz (block (result i32) (br 1 (i32.const 0x8))))))
    )
    (local.get 0)
  )
)

```

File: wasm-text.md [hunk 3]
```diff
@@ -207,29 +230,46 @@ In the text format, it is also allowed to have a conditional without the `else`
...
 Intitial memory data, and initial table elements can be given inline in the text format.
+We supply macros that will unfold these definitions when they are part of a body of definitions, such as in a module.
+This is to ensure that the unfolding happens before the different elements in a module are grouped together, so as to maintain their order.
+
+We also supply rules for when the inlined definitions are encountered on top of the `<k>` cell, so that they can be desugared on the fly.
+
+This is useful when specifying modules in the more lax KWasm format, where they can be declared as they are needed.
 
 ```k
+    syntax Identifier ::= "#memId" | "#tabId"
```
Comment (ehildenb) [inline wasm-text.md]:
```suggestion
    syntax Identifier ::= "#memId" | "#tabId"
 // -----------------------------------------
```

An explanation of what these do would also be nice.

Comment (ehildenb) [reply wasm-text.md]: If they're just used when we need an anonymous table identifier or memory identifier, I would suggest `.Identifier`, or `.MemoryIdentifier` and `.TableIdentifier`.

Comment (hjorthjort) [reply wasm-text.md]: Good idea, thanks.

**Comment by hjorthjort** [comment]
@ehildenb Can you have another look at this and see if we can get it merged?

**Comment by ehildenb** [comment]
If you have to change a bunch to get this passing, please ask for re-review.

**Pull request merged**